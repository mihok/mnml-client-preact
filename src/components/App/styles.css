/* Application Component */

@import url('https://fonts.googleapis.com/css?family=Open+Sans:400,700');

.mnml {
  font-size: 12px;
  font-family: 'Open Sans', Arial;
}

.mnml__container {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
}


/* Messenger Theme */

.mnml--messenger {
  position: fixed;
  display: flex;
  flex-direction: column;
  bottom: 0;
  right: 100px;
  width: 280px;
  height: 360px;
}


/* Float Theme */

