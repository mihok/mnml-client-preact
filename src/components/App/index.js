import { h, render, Component } from 'preact';
import './styles.css';
import Chat from '../Chat/';
import ChatBubble from '../ChatBubble/';
import dummy from '../../utils/dummy.js';
import ThemeProvider from '../ThemeProvider/index';
/* import io from "socket.io-client";*/
// NOTE: SOCKET IS LOADED IN THE EXAMPLE SITE BECASUE ROLLUP IS FAILING AT IMPORTING IO
// SAD sAD FAIL SAD ADAAA SAD

const socketPath = 'http://localhost:8000';

class App extends Component {
  state = {
    chatOpen: true,
    messages: dummy(4, 5).messages,
    textBox: '',
    theme: 'Messenger', // wrapped with theme provider + HOC
  };

  componentDidMount () {
    window.jam = this;
    this.socket = io.connect(socketPath, {
      reconnectionAttempts: 10,
      query: 'type=client',
    });

    this.socket.on('operator:message', this.recieveMessage);
  }

  // ====================================
  // Event Handlers
  // ====================================

  toggleChat = (bool) => {
    this.setState({ chatOpen: bool });
  };

  handleInput = (e) => {
    this.setState({ textBox: e.target.value });
  };

  // ====================================
  //  Message Methods
  // ====================================

  // TODO: combine last message if it's by the same author
  sendMessage = (e) => {
    e.preventDefault();

    if (this.state.textBox === '') return;

    this.socket.emit('client:message', this.state.textBox);

    this.setState({
      messages: [...this.state.messages, { content: [this.state.textBox], author: 'client' }],
      textBox: '',
    });
  };

  recieveMessage = (data) => {
    this.setState({
      messages: [...this.state.messages, { content: [data], author: 'operator' }],
    });
  };


  // if the last message was from the same author, combine 'em.
  combineLastMessage = (msg) => {
    if (msg.author !== previousMsg.author) return;
    const messages = this.state.messages;
    const lastMessage = messages[messages[messages.length - 1]];

    previousMsg.content.concat(msg.content);
    return previousMsg;
  };

  // -- Render Methods -> Component Pieces -- //

  renderClosedChat = () => <ChatBubble toggleChat={this.toggleChat} />;

  renderOpenChat = () => (
    <Chat
      toggleChat={this.toggleChat}
      messages={this.state.messages}
      textBox={this.state.textBox}
      handleInput={this.handleInput}
      sendMessage={this.sendMessage}
    />
  );

  renderChat = () =>
    this.state.chatOpen ? this.renderOpenChat() : this.renderClosedChat();

  // -- Component Return -- //

  render () {
    const { chatStyle, chatOpen, theme } = this.state;

    return (
      <ThemeProvider theme={this.state.theme}>
        <div className={`App App__${theme}`}>
          {this.renderChat()}
        </div>
      </ThemeProvider>
    );
  }
}

export default App;
