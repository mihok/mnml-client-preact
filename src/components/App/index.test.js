/* const render = require('preact-render-to-string');
 * const App = require('../thing');*/
import render from 'preact-render-to-string';
import { h } from 'preact';
import App from '.';

describe('App component', () => {
  it('should render ', () => {
    const tree = render(<App />);
    expect(tree).toMatchSnapshot();
  });
});
