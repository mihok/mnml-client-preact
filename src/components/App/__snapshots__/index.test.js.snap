// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render  1`] = `"<div class=\\"App App__Messenger\\"><section class=\\"Chat__Messenger\\"><header class=\\"Chat__Header__Messenger\\"><div class=\\"Chat__OperatorName__Messenger\\">Chat with John</div><button class=\\"Chat__CloseBtn__Messenger\\">Ã—</button></header><ul class=\\"Chat__Body__Messenger\\"></ul><form class=\\"Input__Form__Messenger\\"><input class=\\"Input__Messenger\\" placeholder=\\"Type Here\\" name=\\"messages\\" value /></form></section></div>"`;
